pddl/parser/types_index.py

```diff
@@ -0,0 +1,209 @@
+# license that can be found in the LICENSE file or at
+# https://opensource.org/licenses/MIT.
+#
+
+"""Utility to handle typed lists."""
+import itertools
+from collections import OrderedDict
+from typing import Dict, List, Optional
+from typing import OrderedDict as OrderedDictType
+from typing import Set, Union
+
+from pddl._validation import _check_names_not_a_keyword, _check_types_not_a_keyword
+from pddl.custom_types import name
+from pddl.helpers.base import check, safe_index
+from pddl.parser.symbols import Symbols
+
+
+class TypesIndex:
+    """
+    An index for PDDL types and PDDL names/variables.
+
+    This class is used to index PDDL names and variables by their types.
+    OrderedDict is used to preserve the order of the types and the names, e.g. for predicate variables.
+
+    Other types of validations are performed to ensure that the types index is consistent.
+    """
+
+    def __init__(self) -> None:
+        """Initialize the types index."""
+        self._types_to_items: OrderedDictType[name, Set[name]] = OrderedDict()
+        self._item_to_types: OrderedDictType[name, Set[name]] = OrderedDict()
+
+    def add_item(self, item_name: name, type_tags: Set[name]) -> None:
+        """
+        Add an item to the types index with the given type tags.
+
+        Both the item name and the type tags are validated according to the name type regular expression.
+
+        :param item_name: the item name
+        :param type_tags: the types for the item
+        """
+        _check_names_not_a_keyword({item_name})
+        _check_types_not_a_keyword(type_tags)
+        self._check_item_name_already_present(item_name)
+        self._check_tags_already_present(item_name, type_tags)
+        self._add_item(item_name, type_tags)
+
+    def get_typed_list_of_names(self) -> Dict[name, Optional[name]]:
+        """Get the typed list of names in form of dictionary."""
+        result: Dict[name, Optional[name]] = {}
+        for item, types_tags in self._item_to_types.items():
+            if len(types_tags) > 1:
+                self._raise_multiple_types_error(item, types_tags)
+            type_tag = next(iter(types_tags)) if len(types_tags) == 1 else None
+            result[item] = type_tag
+        return result
+
+    def get_typed_list_of_variables(self) -> OrderedDictType[name, Set[name]]:
+        """Get the typed list of variables in form of dictionary."""
+        return self._item_to_types
+
+    @classmethod
+    def parse_typed_list(cls, tokens: List[Union[str, List[str]]]) -> "TypesIndex":
+        """
+        Parse typed list.
+
+        This method takes in input a list of tokens as returned by the domain or problem parsers,
+        and returns a TypesIndex object.
+
+        The input list of tokens must have the following format:
+        - if the list is not typed, it is simply a list of names
+        - if the list is typed, the format is: [name_1, ..., name_n], "-", [type_1, ..., type_m], ...
```

Comment:
Shouldn't this be `[name_1, ..., name_n], "-", parent_type, ...` as below?

Reply:
outdated comment, please see the code